### PR TITLE
Make ForbiddenSuppress un-suppressable

### PR DESCRIPTION
**Pull request by mgroth0**
Fixes https://github.com/detekt/detekt/issues/7038

Of the [5 options](https://github.com/detekt/detekt/issues/7038#issuecomment-2027003939) listed by @BraisGabin, this is an implementation of option 4. In the issue, there were basically two votes for option 3 and two votes for option 4. If I added my vote, I would have said any of options 1-4 would work (just not 5, since it doesn't fix this issue). I gave option 4 a shot because it seemed the easiest for me.

This PR gets the job done and resolves the issue, but there are some questionable aspects of the implementation. I would ask about these in particular:

- `isForbiddenSuppress` is possibly not robust in that it checks the class based on the class name. This is because I needed to implement the logic in `core`, which can't do an `is` type check for `ForbiddenSuppress`. One alternative I though of would be to add an additional `suppressable` property to the `Rule` class, but I though that would be bad because this would allow other rules to become unsurpassable outside of user configuration. Given that the only alternative is bad, I think that checking based on classname might ok. At least it is reinforced by the tests in `ForbiddenSuppressSpec`. If the classname or package of `ForbiddenSuppress` was ever refactored, those tests should fail.

- The documentation edit is now a bit wordy and can be confusing. Once we finalize the implementation, maybe it can be re-worded in a simpler way.

- Suppressors now do not work at all on ForbiddenSuppress. This matches the original goal, however the annotation and function suppresser configuration properties are still present in the configuration spec and are valid yaml. Not sure the best way to handle this, but maybe the verification of the yaml could do a special check to ensure that these properties are not defined for ForbiddenSuppress in the yaml file. Alternatively, maybe we could somehow remove the suppresser properties completely for the ForbiddenSuppress configuration, but I am not sure how.

- I tried to run `detekt` and `test` tasks before submitting this PR, but got failures in both that were unrelated to this PR. I read the CONTRIBUTING guide and the first step says to run `./gradlew publishToMavenLocal`, but I don't want to do this because that might affect other projects on my machine. So I'm unsure if that is why my `detekt` and `test` tasks are failing in other ways.

### CODE REVIEW

**Comment by cortinico** [comment]
Thanks for the PR @mgroth0. The CI is still red, can you follow up here?

**Comment by cortinico** [comment]
> @cortinico I don't know how to solve this `println` issue. I tried to pass a reference to the logger (in `ProcessingSettings`) but adding this parameter is a massive change that requires it to be passed in hundreds of places in the project.

You should definitely not use the `println` function as we want to use the logger. 
I would not create an extension function on `Rule` though and just handle the print inside the core logic (where you already have a hold of a logger).

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/Suppressions.kt [hunk 1]
```diff
@@ -1,30 +1,44 @@
...
 
+import io.gitlab.arturbosch.detekt.api.Rule
 import io.gitlab.arturbosch.detekt.api.RuleSet
 import org.jetbrains.kotlin.psi.KtAnnotated
 import org.jetbrains.kotlin.psi.KtAnnotationEntry
 import org.jetbrains.kotlin.psi.KtElement
 import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
 import kotlin.text.RegexOption.IGNORE_CASE
 
+internal fun Rule.isForbiddenSuppress() = this::class.qualifiedName == FORBIDDEN_SUPPRESS_QNAME
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/Suppressions.kt:11]: You can implement this using the ruleDescriptor. The rule name inside it gives you the name of the rule so you can implement the same. This way you don't need to pass a new parameter.

**Comment by BraisGabin** [comment]
I like option 1. And I will probably don't show any warning. Or at least not for now. If we receive a lot of issues about this we could add it but I think that, by default, we don't need the warning.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/AnnotationSuppressor.kt [hunk 2]
```diff
@@ -19,6 +19,9 @@ internal fun annotationSuppressorFactory(rule: Rule, bindingContext: BindingCont
         it.qualifiedNameGlobToRegex()
     }
     return if (annotations.isNotEmpty()) {
+        if (rule.isForbiddenSuppress()) {
+            return null
+        }
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/AnnotationSuppressor.kt:24]:
```suggestion
```

No need. If someone say on the detekt configuration fail that they want to suppress all the errors of `ForbiddenSuppress` on the functions annotated with `@Composable` that's their problem. I agree that adding an `@Suppress("ForbiddenSuppress")` can be "accidental" or could not be cached when doing a review. But if you change the configuration of a rule we should honor that.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/FunctionSuppressor.kt [hunk 3]
```diff
@@ -24,6 +24,9 @@ internal fun functionSuppressorFactory(rule: Rule, bindingContext: BindingContex
     val functionMatchers = rule.config.valueOrDefault("ignoreFunction", emptyList<String>())
         .map(FunctionMatcher::fromFunctionSignature)
     return if (functionMatchers.isNotEmpty()) {
+        if (rule.isForbiddenSuppress()) {
+            return null
+        }
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/FunctionSuppressor.kt:29]:
```suggestion
```

Same as before.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/Suppressions.kt [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
+import io.gitlab.arturbosch.detekt.api.Rule
 import io.gitlab.arturbosch.detekt.api.RuleSet
+import io.gitlab.arturbosch.detekt.core.extractRuleName
 import org.jetbrains.kotlin.psi.KtAnnotated
 import org.jetbrains.kotlin.psi.KtAnnotationEntry
 import org.jetbrains.kotlin.psi.KtElement
 import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
 import kotlin.text.RegexOption.IGNORE_CASE
 
+internal fun Rule.isForbiddenSuppress() = this.ruleName == FORBIDDEN_SUPPRESS_NAME
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/Suppressions.kt:12]:
```suggestion
```

Not used

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/Suppressions.kt [hunk 5]
```diff
@@ -18,13 +24,19 @@ fun KtElement.isSuppressedBy(id: String, aliases: Set<String>, ruleSetId: RuleSe
...
         .mapNotNull { it.getArgumentExpression()?.text }
         .map { it.replace(detektSuppressionPrefixRegex, "") }
         .map { it.replace(QUOTES, "") }
         .any { it in acceptedSuppressionIds }
+
+    if (r && isForbiddenSuppressById(id)) {
+        return false
+    }
+
+    return r
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/suppressors/Suppressions.kt:39]:
```suggestion
    if (isForbiddenSuppressById(id)) return false

    return allAnnotationEntries()
        .filter { it.typeReference?.text in suppressionAnnotations }
        .flatMap { it.valueArguments }
        .mapNotNull { it.getArgumentExpression()?.text }
        .map { it.replace(detektSuppressionPrefixRegex, "") }
        .map { it.replace(QUOTES, "") }
        .any { it in acceptedSuppressionIds }
```

**Review by BraisGabin — APPROVED** [review]

**Comment by BraisGabin** [comment]
Sorry for the slow review. I did some changes (see comments above) and I'm going to merge this directly. If you don't agree with any of my changes we can talk about it on other issue or you can open a new PR with the changes that you will add now. I'm doing this just to get the things moving and merge this fast.
